src/net/fe/fightStage/CombatCalculator.java

```diff
@@ -415,8 +426,66 @@ public static int hitRate(Unit a, Unit d){
 		return a.hit() - d.avoid() + a.getWeapon().triMod(d.getWeapon()) * 15;
 	}
 	
-	public static int hitRate(Unit a, Unit d, Double mod){
-		return ((Double)(a.hit() *  mod)).intValue() - d.avoid() + a.getWeapon().triMod(d.getWeapon()) * 15;
+	public static int critRate(Unit a, Unit d) {
+		return a.crit() - d.dodge();
+	}
+	
+	public static int actualHitRate(RNG rng, Unit a, Unit d) {
+		return rng.actualSuccessRate(hitRate(a, d));
+	}
+	
+	public static int actualCritRate(RNG rng, Unit a, Unit d) {
+		return rng.actualSuccessRate(critRate(a, d));
 	}
 	
+	private static int limit(int min, int max, int val) {
+		if(val < min)
+			return min;
+		if(val > max)
+			return max;
+		return val;
+	}
+	
+	/**
+	 * Immutable class containing the damage, hit and crit of a unit when fighting a specific enemy.
+	 * A value of -1 implies the unit can't attack the enemy.
+	 * @author wellme
+	 */
+	public static class MainBattleStats {
+		public final int damage;
+		public final int hit;
+		public final int crit;
+		
+		public MainBattleStats(int damage, int hit, int crit) {
+			this.damage = damage;
+			this.hit = hit;
+			this.crit = crit;
+		}
+		
+		public String formattedDamage() {
+			if(damage < 0)
+				return " --";
+			return String.format("%3d", damage);
+		}
+		
+		public String formattedHit() {
+			if(hit < 0)
+				return " --";
+			if(FEResources.getActualOdds())
+				return String.format("%3d", FEMultiplayer.getSession().getHitRNG().actualSuccessRate(hit));
+			return String.format("%3d", hit);
+		}
+		
+		public String formattedCrit() {
+			if(crit < 0)
+				return " --";
+			if(FEResources.getActualOdds())
+				return String.format("%3d", FEMultiplayer.getSession().getCritRNG().actualSuccessRate(crit));
+			return String.format("%3d", crit);
+		}
+		
+		public boolean canAttack() {
+			return crit < 0;
```

Comment:
Other way around - can attack if `crit >= 0`. This commit disabled the 'x2' preview thing only if the unit is capable of counter-attacking, when we wanted the inverse.